### PR TITLE
Have cmake and scons use same lists of source files

### PR DESCRIPTION
**Pull request by Pentarctagon**
Move lists of source files into their own directory
Make scons use the lists of source files
Move lua build step to src/SConscript and delete src/lua/SConscript
Make cmake use the separate file with the list of lua sources to build

@CelticMinstrel @GregoryLundberg @gfgtdf This is essentially a re-submission of #879 due to me screwing stuff up enough with git while trying to fix something that I decided to just try deleting all the commits, and then resubmitting them.  Unfortunately, a PR with no commits is automatically closed, so I renamed the branch to something more fitting, and made this PR.

### CODE REVIEW

**Comment by gfgtdf** [comment]
>  Unfortunately, a PR with no commits is automatically closed

I think you can reopen prs? By opening a new pr you somwhow 'hide' all previous discussionon on it.

**Comment by Pentarctagon** [comment]
I looked, but I couldn't find any sort of "re-open PR" button, and pushing the above commit to the branch did not show up in the old PR nor did it re-open it.  I did link to the previous PR though, so the discussion isn't exactly hidden.

**Comment by Pentarctagon** [comment]
Is the file `gui/widgets/debug.cpp` still used?  It is compiled by cmake as part of `wesnoth-gui_widget_SRC`, but apparently not by scons.

**Comment by GregoryLundberg** [comment]
My checks from a month or two ago showed gui/widget/debug.cpp was not referenced. I never noticed it was compiled by CMake.

**Comment by CelticMinstrel** [comment]
It seems like something you'd only want in a special debug build. Even if you compiled it, it wouldn't have any effect without the appropriate -D.

**Comment by Pentarctagon** [comment]
So should I keep it as part of the build, or remove it?

**Comment by Pentarctagon** [comment]
I have left `gui/widget/debug.cpp` in the build lists, so now both scons and cmake will build it.

The tools I haven't tried building, since it seems they have their own issues right now as per #874, but copy/pasting:

> source_lists
> src/SConscript
> src/CMakeLists.txt

into an otherwise unmodified and up-to-date local repository results in a successful compilation for both cmake and scons.  The branch itself won't compile though, due to a recent change which, among other things, renamed `src/sdl/color.cpp` and `src/sdl/color.hpp` to `src/color.cpp` and `src/color.hpp`

**Comment by GregoryLundberg** [comment]
Feel free to ignore my PR. If yours causes a merge conflict on any of mine, I'll fix 'em.

The issues are that that either they fail to link or too much dead code is needed to get them to link.

**Comment by Pentarctagon** [comment]
@GregoryLundberg Yeah, git is already reporting this as causing a merge conflict with `src/CMakeLists.txt` and `src/SConscript`, but that was kind of inevitable.  I would have actually just taken your changes to what source files get built and merged them into my changes, since that isn't overly difficult and I would have been able to test building the tools, but your PR also made C++ changes as well, so just changing the lists of source files for the targets wouldn't have helped a whole lot.

**Comment by Pentarctagon** [comment]
So, unless anyone else sees any issues/has any suggestions, I don't have anything else to add to this PR.

File: src/SConscript [hunk 1]
```diff
@@ -6,42 +6,42 @@ from glob import glob
 
 Import("*")
 
+#
+# load list of sources from the file in scons/sources/
```
Comment (CelticMinstrel) [inline src/SConscript]: This comment is inaccurate?

File: src/SConscript [hunk 2]
```diff
@@ -635,118 +154,46 @@ if env["PLATFORM"] == 'darwin':
...
+#---exploder---
+exploder_sources = GetSources("exploder")
+client_env.WesnothProgram("exploder", exploder_sources + [libwesnoth_core, libwesnoth_sdl, libwesnoth], have_client_prereqs, OBJPREFIX = "exploder_", LIBS = ["$LIBS", "png"])
+
+#---test---
+test_sources = GetSources("test")
+test = test_env.WesnothProgram("test", test_sources +  [libwesnoth_extras, libwesnoth_core, libwesnoth, libwesnoth_sdl, libwesnoth_extras], have_test_prereqs)
+#---end valid scons targets---#
+
+#---start invalid scons targets---#
```
Comment (CelticMinstrel) [inline src/SConscript]: What do you mean by "invalid" here...

**Comment by CelticMinstrel** [comment]
So, I think this is a good idea and should be merged, but the conflicts need to be resolved. Presumably that's not too difficult to do? (Most likely they're from added/removed/moved files since the PR was opened.)

**Comment by GregoryLundberg** [comment]
hold off until I fix my PR it's looking like this and @Vultraz latest in this area will render it moot but there may be a tweak or two needed here

File: src/SConscript (hunk 2 above)
Comment (Pentarctagon) [inline src/SConscript]: in the root SConstruct, the `default_target` variable is specified as only accepting the values `wesnoth wesnothd campaignd cutter exploder test`.

File: src/SConscript (hunk 1 above)
Comment (Pentarctagon) [inline src/SConscript]: Whoops, that where it was when I first submitted this and it was just for scons.  WIll fix.

**Comment by Pentarctagon** [comment]
@CelticMinstrel The conflicts are, I believe, that the lists of source files in `src/SConscript` and `src/CMakeLists.txt` have changed since I submitted this PR.  I've been keeping this PR up-to-date with those changes though, and squashing those updates together which is why the "Finish changing over to source lists" commit keeps moving.

**Comment by Pentarctagon** [comment]
@CelticMinstrel Also, I suppose I should add, I have no idea how to resolve merge conflicts with git.  I think it's supposed to involve `git mergetool`, but that keeps telling me that "No files need merging".

**Comment by Vultraz** [comment]
One needs to rebase the branch. You resolve conflicts while doing that. However, GitHub now has a Resolve Conflicts button.

**Comment by CelticMinstrel** [comment]
Just to be sure, there have been no CMake/scons changes since this PR was submitted, right? Other than the source lists.

**Comment by Pentarctagon** [comment]
Alright, I'm going to double-check a few things, but I think I managed to fix the merge conflicts without horribly breaking the PR.

@CelticMinstrel With the exception of [this one](https://github.com/wesnoth/wesnoth/commit/9d51c40c5cd2dc60696ab457d1d5e08186d51e2e), yes.  Though in that case I had already removed scons' `libdummy_video` library to bring it in sync with cmake, and the only other changes was the actual removal of `dummy_video.cpp`.

**Comment by CelticMinstrel** [comment]
Okay, just waiting to see what @GregoryLundberg has to say, I guess.

**Pull request merged**